File: spaceshift/g/g.js
package("spaceshift.g", function(g){
	"use strict";

	g.TAU = Math.PI * 2;
	g.R = R;
	function R(low, high){
		return (high - low) * Math.random() + low;
	}

	g.V2 = V2;
	function V2(x, y){
		this.x = x;
		this.y = y;
	};


	V2.prototype = {
		length2: function(){
			return this.x * this.x + this.y * this.y;
		},
		length: function(){
			return Math.sqrt(this.x * this.x + this.y * this.y);
		}
	};

	g.U = function(x) {
		return x < -1 ? -1 : x > 1 ? 1 : x;
	};

	g.U2 = function(v){
		var n = new V2(v.x, v.y);
		var d = v.length();
		n.x /= d;
		n.y /= d;
		return n;
	};
});

File: spaceshift/g/g.go
package g

import (
	"math"
	"math/rand"
)

const Tau = 2 * math.Pi

type V2 struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

func R(low, high float64) float64 {
	return (high-low)*rand.Float64() + low
}

func U(v float64) float64 {
	if v < -1 {
		return -1
	}
	if v > 1 {
		return 1
	}
	return v
}

func (a V2) Neg() V2 { return V2{-a.X, -a.Y} }

func (a V2) Add(b V2) V2        { return V2{a.X + b.X, a.Y + b.Y} }
func (a V2) Sub(b V2) V2        { return V2{a.X - b.X, a.Y - b.Y} }
func (a V2) Scale(s float64) V2 { return V2{a.X * s, a.Y * s} }

func (a V2) Len() float64  { return math.Sqrt(a.X*a.X + a.Y*a.Y) }
func (a V2) Len2() float64 { return a.X*a.X + a.Y*a.Y }

func Dist(a, b V2) float64 { return a.Sub(b).Len() }

func Cos(v float64) float64 { return math.Cos(v) }
func Sin(v float64) float64 { return math.Sin(v) }
